services/src/functional-tests/src/test/groovy/org/openkilda/functionaltests/spec/flows/FlowDiversitySpec.groovy

```diff
@@ -0,0 +1,417 @@
+    }
+
+    def "Diverse flows are built through the same path if there are no alternative paths available"() {
+        given: "Two active neighboring switches with two not overlapping paths at least"
+        def (Switch srcSwitch, Switch dstSwitch) = getSwitchPair(2)
+
+        and: "Create a flow going through these switches"
+        def flow1 = flowHelper.randomFlow(srcSwitch, dstSwitch)
+        flowHelper.addFlow(flow1)
+        def flow1Path = PathHelper.convert(northbound.getFlowPath(flow1.id))
+
+        and: "Make all alternative paths unavailable (bring ports down on the source switch)"
+        List<PathNode> broughtDownPorts = []
+        database.getPaths(srcSwitch.dpId, dstSwitch.dpId)*.path.findAll { it != flow1Path }.unique {
+            it.first()
+        }.each { path ->
+            def src = path.first()
+            broughtDownPorts.add(src)
+            northbound.portDown(src.switchId, src.portNo)
+        }
+        Wrappers.wait(WAIT_OFFSET) {
+            assert northbound.getAllLinks().findAll {
+                it.state == IslChangeType.FAILED
+            }.size() == broughtDownPorts.size() * 2
+        }
+
+        when: "Create the second flow with diversity enabled"
+        def flow2 = flowHelper.randomFlow(srcSwitch, dstSwitch, false, [flow1]).tap { it.diverseFlowId = flow1.id }
+        flowHelper.addFlow(flow2)
+        def flow2Path = PathHelper.convert(northbound.getFlowPath(flow2.id))
+
+        then: "The second flow is built through the same path as the first flow"
+        flow2Path == flow1Path
+
+        and: "Restore topology, delete flows and reset costs"
+        broughtDownPorts.every { northbound.portUp(it.switchId, it.portNo) }
+        [flow1, flow2].each { flowHelper.deleteFlow(it.id) }
+        Wrappers.wait(discoveryInterval + WAIT_OFFSET) {
+            northbound.getAllLinks().each { assert it.state != IslChangeType.FAILED }
+        }
+        database.resetCosts()
+    }
+
+    def "Links and switches get extra cost that is considered when calculating diverse flow paths"() {
+        given: "Two active neighboring switches with three not overlapping paths at least"
+        def (Switch srcSwitch, Switch dstSwitch) = getSwitchPair(3)
+
+        and: "Create a flow going through these switches"
+        def flow1 = flowHelper.randomFlow(srcSwitch, dstSwitch)
+        flowHelper.addFlow(flow1)
+        def flow1Path = PathHelper.convert(northbound.getFlowPath(flow1.id))
+
+        and: "Make all min alternative paths less preferable by 1 for diverse flows"
+        def altPaths = database.getPaths(srcSwitch.dpId, dstSwitch.dpId)*.path
+        altPaths.remove(flow1Path)
+        def flow1PathCost = pathHelper.getCost(flow1Path) + diversityIslWeight + diversitySwitchWeight * 2
+        def minAltPathCost = pathHelper.getCost(altPaths.min { it.size() }) + diversitySwitchWeight * 2
+        if (flow1PathCost > minAltPathCost) {
```

Comment:
Ok, but we should understand that if `flow1PathCost <= minAltPathCost` then we will not test the `+1` boundary cost difference (which is explicitly stated by step description). You could have just removed the `if`, then you'll get the negative difference and it will kind of work, but only while `islCost` is bigger than the negative `difference` (L187 may get a negative cost). Just commenting so that all of us understand it.
I would opt for eliminating the conditional branching and always go for the expected +1 difference between current and alt path. Right now the test has multiple different ways of being executed